hardware/VirtualThermostat.cpp

```diff
@@ -0,0 +1,740 @@
+#include "stdafx.h"
+#include "../main/Helper.h"
+#include "../main/Logger.h"
+#include "../main/SQLHelper.h"
+#include "../main/RFXtrx.h"
+
+#include "VirtualThermostat.h"
+
+#include "../main/localtime_r.h"
+#include "../main/mainworker.h"
+#include "../notifications/NotificationHelper.h"
+
+#include <map>
+
+std::string GetDeviceValue(const char * FieldName, const char *Idx)
+{
+	TSqlQueryResult result = m_sql.safe_query("SELECT %s from DeviceStatus WHERE (ID == %s )", FieldName, Idx);
+	if (!result.empty())
+		return  result[0][0];
+	else
+		return  "";
+}
+
+void CircularBuffer::Clear()
+{
+	for (int i = 0; i<Size; i++)Value[i] = 0;
+	index = 0;
+	Sum = 0;
+}
+CircularBuffer::CircularBuffer(int pSize)
+{
+	Value = new double[pSize];
+	Size = pSize;
+	Clear();
+}
+CircularBuffer::~CircularBuffer()
+{
+	delete[] Value;
+}
+int CircularBuffer::GetNext()
+{
+	if (index >= (Size - 1))
+		return 0;
+	else
+		return index + 1;
+}
+double CircularBuffer::Put(double val)
+{
+	double lastv = Value[index];
+	Value[index] = val;
+	index = GetNext();
+	Sum -= lastv;
+	Sum += val;
+	return lastv;
+}
+double CircularBuffer::GetLast()
+{
+	//return last recorded value
+	return Value[index];
+}
+
+double CircularBuffer::GetSum()
+{
+	return Sum;
+}
+
+LastValue::LastValue(float pdelta)
+{
+	Delta = pdelta;
+}
+double LastValue::Get(int index)
+{
+	return LastValues[index];
+}
+void   LastValue::Put(int index, double value)
+{
+	LastValues[index] = value;
+}
+bool LastValue::AsChanged(int index, double value)
+{
+	return AsChanged(index, value, Delta);
+}
+bool LastValue::AsChanged(int index, double value, double delta)
+{
+	if (fabs(LastValues[index] - value)  > delta)
+	{
+		LastValues[index] = value;
+		return true;
+	}
+	return false;
+}
+
+std::string ToString(double value, const char * format = "%f")
+{
+	char buf[64];
+	snprintf(buf, sizeof(buf) - 2, format, value);
+	return buf;
+}
+
+std::string ToString(int value, const char * format = "%d")
+{
+	char buf[64];
+	snprintf(buf, sizeof(buf) - 2, format, value);
+	return buf;
+}
+
+
+//option management
+void getOption(TOptionMap &Option, const char * optionName, int & value)
+{
+	value = atoi(Option[optionName].c_str());
```

Comment:
maybe better to use the c++ function std::stoi, std::stol,....

Reply:
ok done